src/nvim/ops.c

```diff
@@ -2637,12 +2637,86 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
    * special characters (newlines, etc.).
    */
   if (regname == '.') {
-    (void)stuff_inserted((dir == FORWARD ? (count == -1 ? 'o' : 'a') :
-                          (count == -1 ? 'O' : 'i')), count, FALSE);
-    /* Putting the text is done later, so can't really move the cursor to
-     * the next character.  Use "l" to simulate it. */
-    if ((flags & PUT_CURSEND) && gchar_cursor() != NUL)
-      stuffcharReadbuff('l');
+    bool non_linewise_vis = (VIsual_active && VIsual_mode != 'V');
+
+    // PUT_LINE has special handling below which means we use 'i' to start.
+    char command_start_char = non_linewise_vis ? 'c' :
+      (flags & PUT_LINE ? 'i' : (dir == FORWARD ? 'a' : 'i'));
+
+    // To avoid being the affect of 'autoindent' on linewise puts, we create a
+    // new line with `:put _`.
+    if (flags & PUT_LINE) {
+      do_put('_', NULL, dir, 1, PUT_LINE);
+    }
+
+    // If given a count when pasting linewise, we stuff the readbuf with the
+    // dot register 'count' times split by newlines.
+    if (flags & PUT_LINE) {
+      stuffcharReadbuff(command_start_char);
+      for (; count > 0; count--) {
+        (void)stuff_inserted(NUL, 1, count != 1);
+        if (count != 1) {
+          // In order to avoid 'autoindent' affecting us if the text from the
+          // previous insert begins with whitespace, we use C-u to remove all
+          // whitespace at the beginning of the next line.
+          // We can't just insert Ctrl_U into readbuf1, because with
+          // 'backspace' including "eol" this would go back to the previous
+          // line if 'autoindent' is not on, hence we insert a dummy space just
+          // so that it can be removed with the C-u.
```

Comment:
suggestion for brevity:

> To avoid 'autoindent' affecting the text, use Ctrl_U to remove any whitespace. Can't just insert Ctrl_U into readbuf1, this would go back to the previous line in the case of 'noautoindent' and 'backspace' includes "eol". So we insert a dummy space for Ctrl_U to consume.